File: scala/saddle-points/src/main/scala/SaddlePoints.scala
case class Matrix(rows: Seq[Seq[Int]]) {

  case class MatrixValue(row: Int, col: Int, value: Int)

  val matrix: Seq[MatrixValue] =
    rows
      .zipWithIndex
      .flatMap { case (values, row) =>
        values
          .zipWithIndex
          .map { case (n, col) => MatrixValue(row, col, n) }
      }

  def saddlePoints: Set[(Int, Int)] = {
    val rowCandidates = select(_.row, row => row.maxBy(_.value))
    val colCandidates = select(_.col, col => col.minBy(_.value))

    rowCandidates.intersect(colCandidates).map(v => (v.row, v.col))
  }

  private def select(grouping: MatrixValue => Int, selector: Seq[MatrixValue] => MatrixValue) =
    matrix
      .groupBy(grouping)
      .values
      .flatMap { group =>
        val target = selector.apply(group)
        group.filter(_.value == target.value)
      }
      .toSet
}

File: ruby/saddle-points/saddle_points.rb
class Matrix
  attr_reader :rows, :columns, :saddle_points

  def initialize(data)
    @rows = data.lines.map { |row| row.split.map(&:to_i) }
    @columns = rows.transpose
    @saddle_points = []

    (0..@rows.length - 1).each do |row|
      (0..@columns.length - 1).each do |col|
        value = @rows[row][col]
        @saddle_points.append([row, col]) if @rows[row].all? { |v| v <= value } && @columns[col].all? { |v| v >= value }
      end
    end
  end
end
